File: golang/sql/sql.go
package postgres

import (
	"database/sql"
	"fmt"
	"os"
	"strconv"

	_ "github.com/lib/pq"
)

type Postgres struct {
	host     string
	port     int
	user     string
	password string
	database string
}

var Client *sql.DB

func (p *Postgres) Connect() {

	port, _ := strconv.Atoi(os.Getenv("DB_PORT"))
	credentials := Postgres{
		host:     os.Getenv("DB_HOST"),
		port:     port,
		user:     os.Getenv("DB_USER"),
		password: os.Getenv("DB_PASSWORD"),
		database: os.Getenv("DB_NAME"),
	}

	connStr := fmt.Sprintf("host=%s port=%v user=%s password=%s dbname=%s sslmode=disable", credentials.host, credentials.port, credentials.user, credentials.password, credentials.database)

	db, err := sql.Open("postgres", connStr)

	if err != nil {
		panic(err)
	}

	err = db.Ping()

	if err != nil {
		panic(err)
	}

	fmt.Println("POSTGRES CONNECTION SUCCESS!")

	Client = db
}

File: node/src/sql/sql.ts
import { Client } from "pg";

export interface ISQL {
    connect() : Promise<Client>;
    runQuery(query: string): Promise<any>;
    disconnect() : Promise<void>;
}

export interface ISQLCredentials {
    port: number;
    host: string;
    user: string;
    password: string;
    database?: string;
    schema?: string;
}

export class SQL implements ISQL {
    private static instance: Client;

    async connect(): Promise<Client> {

        const credentials : ISQLCredentials = {
            port: parseInt(process.env.DB_PORT!),
            host: process.env.DB_HOST!,
            user: process.env.DB_USER!,
            password: process.env.DB_PASSWORD!,
            database: process.env.DB_NAME,
        }

        try {
            if (!SQL.instance) {
                
                SQL.instance = new Client({
                    user: credentials.user,
                    password: credentials.password,
                    host: credentials.host,
                    port: credentials.port,
                    database: credentials.database,
                });
    
                await SQL.instance.connect();
    
                console.log("SQL CONNECTION SUCCESS!");
            }

            return SQL.instance;
        } catch (err) {
            const message = `SQL CONNECTION ERROR\nCREDENTIALS:\n${JSON.stringify(credentials)}ERROR:\n${err}`;
            console.log(message);
            throw new Error(message);
        }
    }

    async disconnect(): Promise<void> {
        try {
            await SQL.instance.end();
            console.log("SQL DISCONNECT!");

        } catch (err) {
            const message = `SQL DISCONNECTION ERROR\nERROR:\n${err}`;
            console.log(message);
            throw new Error(message);
        }
    }

    async runQuery(query: string): Promise<any> {
        try {
            const res = await SQL.instance?.query(query);
            return res;
        } catch(err) {
            const message = `SQL QUERY ERROR\nQUERY:\n${query}ERROR:\n${err}`;
            console.log(message);
            throw new Error(message);
        }
    }
}

export default SQL;